### PR TITLE
Allow device policy execution for IntersectionShaper, DistributedClosestPoint

### PR DESCRIPTION
**Pull request by bmhan12**
This PR:

- Contains work related to issue #1339 
- Refactors IntersectionShaper and DistributedClosestPoint to allow for execution with a device policy (previously required unified policy to run on device)

### CODE REVIEW

File: src/axom/quest/detail/DistributedClosestPointImpl.hpp [hunk 1]
```diff
@@ -1062,8 +1062,14 @@ class DistributedClosestPointExec : public DistributedClosestPointImpl
         auto it = m_bvh->getTraverser();
         const int rank = m_rank;
 
-        double* sqDistThresh = axom::allocate<double>(1, m_allocatorID);
-        *sqDistThresh = m_sqDistanceThreshold;
+        axom::Array<double> sqDistThresh_host(
+          1,
+          1,
+          axom::execution_space<axom::SEQ_EXEC>::allocatorID());
```
Comment (kennyweiss) [inline src/axom/quest/detail/DistributedClosestPointImpl.hpp:1068]: 👍

File: src/axom/quest/IntersectionShaper.hpp [hunk 2]
```diff
@@ -639,40 +628,39 @@ class IntersectionShaper : public Shaper
...
     constexpr int NUM_COMPS_PER_VERT = 3;
     constexpr int NUM_TETS_PER_HEX = 24;
     constexpr double ZERO_THRESHOLD = 1.e-10;
 
-    // Save current/default allocator
-    const int current_allocator = axom::getDefaultAllocatorID();
-
-    // Determine new allocator (for CUDA/HIP policy, set to Unified)
-    // Set new default to device
-    axom::setDefaultAllocator(::getUmpireDeviceId<ExecSpace>());
```
Comment (kennyweiss) [inline src/axom/quest/IntersectionShaper.hpp:652]: 👍
+1 for reducing our usage of the global default allocator!

**Review by kennyweiss — APPROVED** [review]
Thanks for following up on this @bmhan12 !

Did you notice any performance improvements in this branch w.r.t the unified memory usage in `develop`?

File: src/axom/quest/IntersectionShaper.hpp [hunk 3]
```diff
@@ -705,19 +693,20 @@ class IntersectionShaper : public Shaper
     this->getDC()->RegisterField(volFracName, volFrac);
 
     // Initialize hexahedral elements
-    m_hexes = axom::Array<HexahedronType>(NE, NE);
-    axom::ArrayView<HexahedronType> hexes_view = m_hexes.view();
+    m_hexes = axom::Array<HexahedronType>(NE, NE, kernel_allocator);
```
Comment (white238) [inline src/axom/quest/IntersectionShaper.hpp:696]: Should we be toggling between `kernel` and `device` in our naming?

**Comment by bmhan12** [comment]
> Did you notice any performance improvements in this branch w.r.t the unified memory usage in develop?

I didn't do any major benchmarking, but at least for the `quest_intersection_shaper` unit tests, I was seeing:

- On CUDA, about the same overall runtime  (maybe it would be different if I disabled blueos's ats?)
- On HIP, the unit test completion time goes from about an average runtime of 11.5 seconds to 9 seconds.

File: src/axom/quest/IntersectionShaper.hpp (hunk 3 above)
Comment (bmhan12) [reply src/axom/quest/IntersectionShaper.hpp]: Good catch! Changed `kernel_allocator` to `device_allocator` to keep naming standard.

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**